README.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


  - [ledger-app-cosmos](#ledger-app-cosmos)
  - [Installation](#installation)
  - [Demo](#demo)
  - [Documentation](#documentation)
  - [Screenshot](#screenshot)
- [Classes](#classes)
  - [Class: Cosmos](#class-cosmos)
    - [Table of contents](#table-of-contents)
    - [Constructors](#constructors)
    - [Properties](#properties)
    - [Accessors](#accessors)
    - [Methods](#methods)
  - [Class: Ledger](#class-ledger)
    - [Hierarchy](#hierarchy)
    - [Table of contents](#table-of-contents-1)
    - [Constructors](#constructors-1)
    - [Properties](#properties-1)
    - [Accessors](#accessors-1)
    - [Methods](#methods-1)
- [Enums](#enums)
  - [Enumeration: EVENTS](#enumeration-events)
    - [Table of contents](#table-of-contents-2)
    - [Enumeration members](#enumeration-members)
- [ledger-app-cosmos](#ledger-app-cosmos-1)
  - [Table of contents](#table-of-contents-3)
    - [Enumerations](#enumerations)
    - [Classes](#classes-1)
    - [Type aliases](#type-aliases)
    - [Variables](#variables)
  - [Type aliases](#type-aliases-1)
    - [APK](#apk)
    - [ApplicationInformation](#applicationinformation)
    - [Device](#device)
    - [DeviceInformation](#deviceinformation)
    - [Exception](#exception)
    - [LedgerOptions](#ledgeroptions)
    - [Model](#model)
    - [PK](#pk)
    - [Signature](#signature)
    - [Version](#version)
  - [Variables](#variables-1)
    - [Ledgerer](#ledgerer)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->


<a name="readmemd"></a>

ledger-app-cosmos / [Exports](#modulesmd)

## ledger-app-cosmos

#### TypeScript/JavaScript client library to communicate with a Tendermint/Cosmos App running in a Ledger Nano S/X

## Installation

Install with [npm](https://www.npmjs.com):

```sh
$ npm install --save ledger-app-cosmos
```

Install with [yarn](https://yarnpkg.com):

```sh
$ npm run add ledger-app-cosmos
```

## [Demo](https://ledger-app-cosmos.netlify.app/)

#### [Try online demo](https://ledger-app-cosmos.netlify.app/)

## [Documentation](https://www.npmjs.com/package/ledger-app-cosmos)

#### [Try online documentation](https://www.npmjs.com/package/ledger-app-cosmos)

## Screenshot

![Example](./documentation/media/screenshot.png)

# Classes


<a name="classescosmosmd"></a>

[ledger-app-cosmos](#readmemd) / [Exports](#modulesmd) / Cosmos

## Class: Cosmos

[[include:cosmos/properties.md]]

[[include:installation.md]]

[[include:cosmos/usage.md]]

[[include:cosmos/demo.md]]

### Table of contents

#### Constructors

- [constructor](#constructor)

#### Properties

- [scrambleKey](#scramblekey)

#### Accessors

- [transport](#transport)

#### Methods

- [debug](#debug)
- [getAddressAndPublicKey](#getaddressandpublickey)
- [getApplicationInformation](#getapplicationinformation)
- [getBech32FromPK](#getbech32frompk)
- [getDeviceInformation](#getdeviceinformation)
- [getPublicKey](#getpublickey)
- [getVersion](#getversion)
- [serializeHRP](#serializehrp)
- [serializePath](#serializepath)
- [showAddressAndPubKey](#showaddressandpubkey)
- [sign](#sign)

### Constructors

#### constructor

• **new Cosmos**(`transport`, `scrambleKey?`)

##### Parameters

| Name | Type |
| :------ | :------ |
| `transport` | `default` |
| `scrambleKey` | `string` |

##### Defined in

[src/lib/cosmos/index.ts:71](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/cosmos/index.ts#L71)

### Properties

#### scrambleKey

• `Readonly` **scrambleKey**: `string`

### Accessors

#### transport

• `get` **transport**(): `default`

##### Returns

`default`

##### Defined in

[src/lib/cosmos/index.ts:94](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/cosmos/index.ts#L94)

• `set` **transport**(`transport`): `void`

##### Parameters

| Name | Type |
| :------ | :------ |
| `transport` | `default` |

##### Returns

`void`

##### Defined in

[src/lib/cosmos/index.ts:98](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/cosmos/index.ts#L98)

### Methods

#### debug

▸ **debug**(`debug`): `void`

##### Parameters

| Name | Type |
| :------ | :------ |
| `debug` | `boolean` |

##### Returns

`void`

##### Defined in

[src/lib/cosmos/index.ts:82](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/cosmos/index.ts#L82)

___

#### getAddressAndPublicKey

▸ **getAddressAndPublicKey**(`path`, `hrp`, `p1?`, `p2?`): `Promise`<[`APK`](#apk)\>

##### Parameters

| Name | Type | Default value |
| :------ | :------ | :------ |
| `path` | `string` \| `number`[] | `undefined` |
| `hrp` | `string` | `undefined` |
| `p1` | `number` | `undefined` |
| `p2` | `number` | `0` |

##### Returns

`Promise`<[`APK`](#apk)\>

##### Defined in

[src/lib/cosmos/index.ts:367](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/cosmos/index.ts#L367)

___

#### getApplicationInformation

▸ **getApplicationInformation**(): `Promise`<[`ApplicationInformation`](#applicationinformation)\>

##### Returns

`Promise`<[`ApplicationInformation`](#applicationinformation)\>

##### Defined in

[src/lib/cosmos/index.ts:190](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/cosmos/index.ts#L190)

___

#### getBech32FromPK

▸ **getBech32FromPK**(`hrp`, `pk`): `string`

##### Parameters

| Name | Type |
| :------ | :------ |
| `hrp` | `string` |
| `pk` | `Buffer` |

##### Returns

`string`

##### Defined in

[src/lib/cosmos/index.ts:143](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/cosmos/index.ts#L143)

___

#### getDeviceInformation

▸ **getDeviceInformation**(): `Promise`<[`DeviceInformation`](#deviceinformation)\>

##### Returns

`Promise`<[`DeviceInformation`](#deviceinformation)\>

##### Defined in

[src/lib/cosmos/index.ts:257](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/cosmos/index.ts#L257)

___

#### getPublicKey

▸ **getPublicKey**(`path`): `Promise`<[`PK`](#pk)\>

##### Parameters

| Name | Type |
| :------ | :------ |
| `path` | `string` \| `number`[] |

##### Returns

`Promise`<[`PK`](#pk)\>

##### Defined in

[src/lib/cosmos/index.ts:339](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/cosmos/index.ts#L339)

___

#### getVersion

▸ **getVersion**(`force?`): `Promise`<[`Version`](#version)\>

##### Parameters

| Name | Type | Default value |
| :------ | :------ | :------ |
| `force` | `boolean` | `true` |

##### Returns

`Promise`<[`Version`](#version)\>

##### Defined in

[src/lib/cosmos/index.ts:174](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/cosmos/index.ts#L174)

___

#### serializeHRP

▸ **serializeHRP**(`hrp`): `Buffer`

##### Parameters

| Name | Type |
| :------ | :------ |
| `hrp` | `string` |

##### Returns

`Buffer`

##### Defined in

[src/lib/cosmos/index.ts:128](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/cosmos/index.ts#L128)

___

#### serializePath

▸ **serializePath**(`path`): `Promise`<`Buffer`\>

##### Parameters

| Name | Type |
| :------ | :------ |
| `path` | `string` \| readonly `number`[] |

##### Returns

`Promise`<`Buffer`\>

##### Defined in

[src/lib/cosmos/index.ts:157](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/cosmos/index.ts#L157)

___

#### showAddressAndPubKey

▸ **showAddressAndPubKey**(`path`, `hrp`, `p1?`, `p2?`): `Promise`<[`APK`](#apk)\>

##### Parameters

| Name | Type | Default value |
| :------ | :------ | :------ |
| `path` | `string` \| `number`[] | `undefined` |
| `hrp` | `string` | `undefined` |
| `p1` | `number` | `undefined` |
| `p2` | `number` | `0` |

##### Returns

`Promise`<[`APK`](#apk)\>

##### Defined in

[src/lib/cosmos/index.ts:405](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/cosmos/index.ts#L405)

___

#### sign

▸ **sign**(`path`, `data`, `size?`): `Promise`<[`Signature`](#signature)\>

##### Parameters

| Name | Type |
| :------ | :------ |
| `path` | `string` \| `number`[] |
| `data` | `any` |
| `size` | `number` |

##### Returns

`Promise`<[`Signature`](#signature)\>

##### Defined in

[src/lib/cosmos/index.ts:414](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/cosmos/index.ts#L414)


<a name="classesledgermd"></a>

[ledger-app-cosmos](#readmemd) / [Exports](#modulesmd) / Ledger

## Class: Ledger

[[include:ledgerer/properties.md]]

[[include:installation.md]]

[[include:ledgerer/usage.md]]

[[include:ledgerer/demo.md]]

### Hierarchy

- `EventEmitter`

  ↳ **`Ledger`**

### Table of contents

#### Constructors

- [constructor](#constructor)

#### Properties

- [EVENTS](#events)
- [error](#error)
- [isConnecting](#isconnecting)
- [isDisconnecting](#isdisconnecting)
- [EVENTS](#events)
- [captureRejectionSymbol](#capturerejectionsymbol)
- [captureRejections](#capturerejections)
- [defaultMaxListeners](#defaultmaxlisteners)
- [errorMonitor](#errormonitor)

#### Accessors

- [cosmos](#cosmos)
- [device](#device)
- [model](#model)
- [transport](#transport)
- [instance](#instance)

#### Methods

- [addListener](#addlistener)
- [connect](#connect)
- [debug](#debug)
- [disconnect](#disconnect)
- [emit](#emit)
- [eventNames](#eventnames)
- [getAddressAndPublicKey](#getaddressandpublickey)
- [getApplicationInformation](#getapplicationinformation)
- [getBech32FromPK](#getbech32frompk)
- [getDeviceInformation](#getdeviceinformation)
- [getMaxListeners](#getmaxlisteners)
- [getPublicKey](#getpublickey)
- [getVersion](#getversion)
- [listenerCount](#listenercount)
- [listeners](#listeners)
- [off](#off)
- [on](#on)
- [once](#once)
- [prependListener](#prependlistener)
- [prependOnceListener](#prependoncelistener)
- [rawListeners](#rawlisteners)
- [reconnect](#reconnect)
- [removeAllListeners](#removealllisteners)
- [removeListener](#removelistener)
- [serializeHRP](#serializehrp)
- [serializePath](#serializepath)
- [setMaxListeners](#setmaxlisteners)
- [showAddressAndPubKey](#showaddressandpubkey)
- [sign](#sign)
- [transporter](#transporter)
- [u2f](#u2f)
- [webBle](#webble)
- [webHID](#webhid)
- [webUSB](#webusb)
- [getEventListeners](#geteventlisteners)
- [listenerCount](#listenercount)
- [on](#on)
- [once](#once)

### Constructors

#### constructor

• **new Ledger**(`options?`)

##### Parameters

| Name | Type |
| :------ | :------ |
| `options` | [`LedgerOptions`](#ledgeroptions) |

##### Overrides

events.EventEmitter.constructor

##### Defined in

[src/lib/Ledger.ts:123](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/Ledger.ts#L123)

### Properties

#### EVENTS

• `Readonly` **EVENTS**: typeof [`EVENTS`](#enumseventsmd)

##### Defined in

[src/lib/Ledger.ts:91](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/Ledger.ts#L91)

___

#### error

• **error**: ``null`` \| `Error` = `null`

##### Defined in

[src/lib/Ledger.ts:90](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/Ledger.ts#L90)

___

#### isConnecting

• **isConnecting**: `boolean` = `false`

##### Defined in

[src/lib/Ledger.ts:93](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/Ledger.ts#L93)

___

#### isDisconnecting

• **isDisconnecting**: `boolean` = `false`

##### Defined in

[src/lib/Ledger.ts:94](https://github.com/idimetrix/ledger-app-cosmos/blob/b96f748/src/lib/Ledger.ts#L94)

___

#### EVENTS

▪ `Static` `Readonly` **EVENTS**: typeof [`EVENTS`](#enumseventsmd)

##### Defined in


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.104](https://github.com/dfinance/ledger/compare/v1.0.103...v1.0.104) (2021-10-04)

### [1.0.103](https://github.com/dfinance/ledger/compare/v1.0.102...v1.0.103) (2021-09-20)

### [1.0.102](https://github.com/dfinance/ledger/compare/v1.0.101...v1.0.102) (2021-09-17)

### [1.0.101](https://github.com/dfinance/ledger/compare/v1.0.100...v1.0.101) (2021-09-17)

### [1.0.100](https://github.com/dfinance/ledger/compare/v1.0.99...v1.0.100) (2021-09-16)

### [1.0.99](https://github.com/dfinance/ledger/compare/v1.0.98...v1.0.99) (2021-09-09)

### [1.0.98](https://github.com/dfinance/ledger/compare/v1.0.97...v1.0.98) (2021-09-09)

### [1.0.97](https://github.com/dfinance/ledger/compare/v1.0.96...v1.0.97) (2021-09-06)

### [1.0.96](https://github.com/dfinance/ledger/compare/v1.0.95...v1.0.96) (2021-09-06)

### [1.0.95](https://github.com/dfinance/ledger/compare/v1.0.94...v1.0.95) (2021-09-06)

### [1.0.94](https://github.com/dfinance/ledger/compare/v1.0.93...v1.0.94) (2021-09-01)

### [1.0.93](https://github.com/dfinance/ledger/compare/v1.0.92...v1.0.93) (2021-09-01)

### [1.0.92](https://github.com/dfinance/ledger/compare/v1.0.91...v1.0.92) (2021-08-31)

### [1.0.91](https://github.com/dfinance/ledger/compare/v1.0.90...v1.0.91) (2021-08-30)

### [1.0.90](https://github.com/dfinance/ledger/compare/v1.0.89...v1.0.90) (2021-08-27)

### [1.0.89](https://github.com/dfinance/ledger/compare/v1.0.88...v1.0.89) (2021-08-25)

### [1.0.88](https://github.com/dfinance/ledger/compare/v1.0.87...v1.0.88) (2021-08-25)

### [1.0.87](https://github.com/dfinance/ledger/compare/v1.0.86...v1.0.87) (2021-08-25)

### [1.0.86](https://github.com/dfinance/ledger/compare/v1.0.85...v1.0.86) (2021-08-24)

### [1.0.85](https://github.com/dfinance/ledger/compare/v1.0.84...v1.0.85) (2021-08-19)

### [1.0.84](https://github.com/dfinance/ledger/compare/v1.0.83...v1.0.84) (2021-08-19)

### [1.0.83](https://github.com/dfinance/ledger/compare/v1.0.82...v1.0.83) (2021-08-19)

### [1.0.82](https://github.com/dfinance/ledger/compare/v1.0.81...v1.0.82) (2021-08-19)

### [1.0.81](https://github.com/dfinance/ledger/compare/v1.0.80...v1.0.81) (2021-08-19)

### [1.0.80](https://github.com/dfinance/ledger/compare/v1.0.79...v1.0.80) (2021-08-19)

### [1.0.79](https://github.com/dfinance/ledger/compare/v1.0.78...v1.0.79) (2021-08-18)

### [1.0.78](https://github.com/dfinance/ledger/compare/v1.0.77...v1.0.78) (2021-08-17)

### [1.0.77](https://github.com/dfinance/ledger/compare/v1.0.76...v1.0.77) (2021-08-17)

### [1.0.76](https://github.com/dfinance/ledger/compare/v1.0.75...v1.0.76) (2021-08-12)

### [1.0.75](https://github.com/dfinance/ledger/compare/v1.0.74...v1.0.75) (2021-08-12)

### [1.0.74](https://github.com/dfinance/ledger/compare/v1.0.73...v1.0.74) (2021-08-10)

### [1.0.73](https://github.com/dfinance/ledger/compare/v1.0.72...v1.0.73) (2021-08-09)

### [1.0.72](https://github.com/dfinance/ledger/compare/v1.0.71...v1.0.72) (2021-08-09)

### [1.0.71](https://github.com/dfinance/ledger/compare/v1.0.70...v1.0.71) (2021-08-07)

### [1.0.70](https://github.com/dfinance/ledger/compare/v1.0.69...v1.0.70) (2021-08-07)

### [1.0.69](https://github.com/dfinance/ledger/compare/v1.0.68...v1.0.69) (2021-08-03)

### [1.0.68](https://github.com/dfinance/ledger/compare/v1.0.67...v1.0.68) (2021-08-02)

### [1.0.67](https://github.com/dfinance/ledger/compare/v1.0.66...v1.0.67) (2021-08-02)

### [1.0.66](https://github.com/dfinance/ledger/compare/v1.0.65...v1.0.66) (2021-08-02)

### [1.0.65](https://github.com/dfinance/ledger/compare/v1.0.64...v1.0.65) (2021-08-02)

### [1.0.64](https://github.com/dfinance/ledger/compare/v1.0.63...v1.0.64) (2021-07-28)

### [1.0.63](https://github.com/dfinance/ledger/compare/v1.0.62...v1.0.63) (2021-07-27)

### [1.0.62](https://github.com/dfinance/ledger/compare/v1.0.61...v1.0.62) (2021-07-27)

### [1.0.61](https://github.com/dfinance/ledger/compare/v1.0.60...v1.0.61) (2021-07-27)

### [1.0.60](https://github.com/dfinance/ledger/compare/v1.0.59...v1.0.60) (2021-07-27)

### [1.0.59](https://github.com/dfinance/ledger/compare/v1.0.58...v1.0.59) (2021-07-27)

### [1.0.58](https://github.com/dfinance/ledger/compare/v1.0.57...v1.0.58) (2021-07-27)

### [1.0.57](https://github.com/dfinance/ledger/compare/v1.0.56...v1.0.57) (2021-07-26)

### [1.0.56](https://github.com/dfinance/ledger/compare/v1.0.55...v1.0.56) (2021-07-26)

### [1.0.55](https://github.com/dfinance/ledger/compare/v1.0.54...v1.0.55) (2021-04-06)

### [1.0.54](https://github.com/dfinance/ledger/compare/v1.0.53...v1.0.54) (2021-03-25)

### [1.0.53](https://github.com/dfinance/ledger/compare/v1.0.52...v1.0.53) (2021-02-03)

### [1.0.52](https://github.com/dfinance/ledger/compare/v1.0.51...v1.0.52) (2021-02-03)

### [1.0.51](https://github.com/dfinance/ledger/compare/v1.0.50...v1.0.51) (2021-01-15)

### [1.0.50](https://github.com/dfinance/ledger/compare/v1.0.49...v1.0.50) (2021-01-10)

### [1.0.49](https://github.com/dfinance/ledger/compare/v1.0.48...v1.0.49) (2020-11-20)

### [1.0.48](https://github.com/dfinance/ledger/compare/v1.0.47...v1.0.48) (2020-11-19)

### [1.0.47](https://github.com/dfinance/ledger/compare/v1.0.46...v1.0.47) (2020-11-16)

### 1.0.46 (2020-11-14)

### [1.0.45](https://github.com/dfinance/ledger/compare/v1.0.44...v1.0.45) (2020-10-16)

### [1.0.44](https://github.com/dfinance/ledger/compare/v1.0.43...v1.0.44) (2020-10-16)

### [1.0.43](https://github.com/dfinance/ledger/compare/v1.0.42...v1.0.43) (2020-10-13)


src/lib/cosmos/index.ts
import crypto from 'crypto';

import Transport from '@ledgerhq/hw-transport';
import { bech32 } from 'bech32';
import Ripemd160 from 'ripemd160';

import { cosmosLog as log } from '../../log';

import {
  APPLICATION,
  CLA,
  ERROR_CODE,
  INS,
  P1_VALUES,
  SIZE,
} from './constants';
import { publicKeyv1, serializePathv1, signSendChunkv1 } from './cosmos1';
import { publicKeyv2, serializePathv2, signSendChunkv2 } from './cosmos2';
import {
  APK,
  ApplicationInformation,
  DeviceInformation,
  PK,
  Signature,
  Version,
} from './types';
import {
  errorCodeToString,
  isArrayOrTypedArray,
  isObject,
  processErrorResponse,
  sortObject,
} from './utils';

const wait: (ms?: number) => Promise<number> = (
  ms: number = 1
): Promise<number> =>
  new Promise(
    (resolve: (value: number | PromiseLike<number>) => void): number =>
      window.setTimeout(resolve, ms)
  );

/**
 *
 * [[include:cosmos/properties.md]]
 *
 * [[include:installation.md]]
 *
 * [[include:cosmos/usage.md]]
 *
 * [[include:cosmos/demo.md]]
 *
 */
export class Cosmos {
  private _transport: Transport = null as any;

  private _version: Version | null = null;
  private _applicationInformation: ApplicationInformation | null = null;
  private _deviceInformation: DeviceInformation | null = null;

  // --- INSTANCE --------------------------------------------------------------

  // private static _instance: Cosmos | null = null;
  //
  // public static get instance(): Cosmos {
  //   return this._instance || (this._instance = new Cosmos());
  // }

  // ---------------------------------------------------------------------------

  constructor(
    transport: Transport,
    public readonly scrambleKey: string = APPLICATION
  ) {
    this.transport = transport;

    log.log('Cosmos created', { transport });
  }

  // --- DEBUG -----------------------------------------------------------------

  public debug(debug: boolean): void {
    log.log('Cosmos:debug', { debug });

    if (debug) {
      log.enableAll();
    } else {
      log.disableAll();
    }
  }

  // --- PUBLIC ----------------------------------------------------------------

  public get transport(): Transport {
    return this._transport;
  }

  public set transport(transport: Transport) {
    log.log('Cosmos:transport', { transport });

    if (!transport) {
      throw new Error('Transport has not been defined');
    }

    this._version = null;
    this._applicationInformation = null;
    this._deviceInformation = null;

    this._transport = transport;

    transport.decorateAppAPIMethods(
      this,
      [
        'getVersion',
        'getApplicationInformation',
        'getDeviceInformation',
        'getPublicKey',
        'getAddressAndPublicKey',
        'showAddressAndPublicKey',
        'sign',
      ],
      this.scrambleKey
    );
  }

  // --- PUBLIC

  public serializeHRP(hrp: string): Buffer {
    log.log('Cosmos:serializeHRP', { hrp });

    if (!hrp || hrp.length < 3 || hrp.length > 83) {
      throw new Error('Invalid HRP');
    }

    const buffer: Buffer = Buffer.alloc(1 + hrp.length);

    buffer.writeUInt8(hrp.length, 0);
    buffer.write(hrp, 1);

    return buffer;
  }

  public getBech32FromPK(hrp: string, pk: Buffer): string {
    log.log('Cosmos:getBech32FromPK', { hrp, pk });

    if (pk.length !== 33) {
      throw new Error('expected compressed public key [31 bytes]');
    }

    const hashSha256: Buffer = crypto.createHash('sha256').update(pk).digest();

    const hashRip: Buffer = new Ripemd160().update(hashSha256).digest();

    return bech32.encode(hrp, bech32.toWords(hashRip));
  }

  public async serializePath(
    path: readonly number[] | string
  ): Promise<Buffer> {
    log.log('Cosmos:serializePath', { path });

    const version: Version = await this.version();

    switch (version.major) {
      case 1:
        return serializePathv1(path);
      case 2:
        return serializePathv2(path);
      default:
        throw version;
    }
  }

  public async getVersion(force: boolean = true): Promise<Version> {
    try {
      log.log('Cosmos:getVersion request');

      const version: Version = await this.version(force);

      log.log('Cosmos:getVersion response', { version });

      return version;
    } catch (error) {
      log.error('Cosmos:getVersion', { error });

      return processErrorResponse(error);
    }
  }

  public async getApplicationInformation(): Promise<ApplicationInformation> {
    if (this?._applicationInformation?.return_code === ERROR_CODE.NoError) {
      return this._applicationInformation;
    }

    try {
      log.log('Cosmos:getApplicationInformation request');

      const response: Uint8Array = await this.send(0xb0, 0x01, 0, 0);

      log.log('Cosmos:getApplicationInformation', { response });

      const codes: Uint8Array = response.slice(-2);
      const code: number = codes[0] * 256 + codes[1];

      if (response[0] !== 1) {
        // Ledger responds with format ID 1. There is no spec for any format != 1
        throw {
          error_message: 'response format ID not recognized',
          return_code: 0x9001,
        };
      }

      const appNameLen: number = response[1];

      const appName: string = Buffer.from(
        response.slice(2, 2 + appNameLen)
      ).toString('ascii');

      let idx: number = 2 + appNameLen;

      const appVersionLen: number = response[idx];

      idx += 1;

      const appVersion: string = Buffer.from(
        response.slice(idx, idx + appVersionLen)
      ).toString('ascii');

      idx += appVersionLen;

      const appFlagsLen: number = response[idx];

      idx += 1;

      const flagLen: number = appFlagsLen;
      const flagsValue: number = response[idx];

      return (this._applicationInformation = {
        appName,
        appVersion,
        error_message: errorCodeToString(code),
        flagLen,
        flag_onboarded: (flagsValue & 4) !== 0,
        flag_pin_validated: (flagsValue & 128) !== 0,
        flag_recovery: (flagsValue & 1) !== 0,
        flag_signed_mcu_code: (flagsValue & 2) !== 0,
        flagsValue,
        return_code: code,
      });
    } catch (error) {
      log.error('Cosmos:getApplicationInformation', { error });

      return processErrorResponse(error);
    }
  }

  public async getDeviceInformation(): Promise<DeviceInformation> {
    if (this?._deviceInformation?.return_code === ERROR_CODE.NoError) {
      return this._deviceInformation;
    }

    try {
      log.log('Cosmos:getDeviceInformation request');

      const response: Uint8Array = await this.send(
        0xe0,
        0x01,
        0,
        0,
        Buffer.from([]),
        [ERROR_CODE.NoError, 0x6e00]
      );

      log.log('Cosmos:getDeviceInformation', { response });

      const codes: Uint8Array = response.slice(-2);
      const code: number = codes[0] * 256 + codes[1];

      if (code === 0x6e00) {
        throw {
          error_message: 'This command is only available in the Dashboard',
          return_code: code,
        };
      }

      const targetId: string = Buffer.from(response.slice(0, 4)).toString(
        'hex'
      );

      let pos: number = 4;

      const secureElementVersionLen: number = response[pos];

      pos += 1;

      const seVersion: string = response
        .slice(pos, pos + secureElementVersionLen)
        .toString();

      pos += secureElementVersionLen;

      const flagsLen: number = response[pos];

      pos += 1;

      const flag: string = Buffer.from(
        response.slice(pos, pos + flagsLen)
      ).toString('hex');

      pos += flagsLen;

      const mcuVersionLen: number = response[pos];
      pos += 1;

      // Patch issue in mcu version
      let tmp: Uint8Array = response.slice(pos, pos + mcuVersionLen);

      if (tmp[mcuVersionLen - 1] === 0) {
        tmp = response.slice(pos, pos + mcuVersionLen - 1);
      }

      const mcuVersion: string = tmp.toString();

      return (this._deviceInformation = {
        error_message: errorCodeToString(code),
        flag,
        mcuVersion,
        return_code: code,
        seVersion,
        targetId,
      });
    } catch (error) {
      log.error('Cosmos:getDeviceInformation', { error });

      return processErrorResponse(error);
    }
  }

  public async getPublicKey(path: number[] | string): Promise<PK> {
    try {
      log.log('Cosmos:getPublicKey request', { path });

      const version: Version = await this.version();

      switch (version.major) {
        case 1:
          return publicKeyv1(this.transport, await this.serializePath(path));
        case 2: {
          return publicKeyv2(
            this.transport,
            Buffer.concat([
              this.serializeHRP('cosmos'),
              await this.serializePath(path),
            ])
          );
        }
        default:
          throw version;
      }
    } catch (error) {
      log.error('Cosmos:getPublicKey', { error });

      return processErrorResponse(error);
    }
  }

  public async getAddressAndPublicKey(
    path: number[] | string,
    hrp: string,
    p1: number = P1_VALUES.ONLY_RETRIEVE,
    p2: number = 0
  ): Promise<APK> {
    try {
      log.log('Cosmos:getAddressAndPublicKey request', { path, hrp, p1, p2 });

      const serializedPath: Buffer = await this.serializePath(path);

      const response: Uint8Array = await this.send(
        CLA,
        INS.GET_ADDR_SECP256K1,
        p1,
        p2,
        Buffer.concat([this.serializeHRP(hrp), serializedPath]),
        [ERROR_CODE.NoError]
      );

      log.log('Cosmos:getAddressAndPublicKey response', { response });

      const codes: Uint8Array = response.slice(-2);
      const code: number = codes[0] * 256 + codes[1];

      return {
        bech32_address: Buffer.from(response.slice(33, -2)).toString(),
        compressed_pk: Buffer.from(response.slice(0, 33)),
        error_message: errorCodeToString(code),
        return_code: code,
      };
    } catch (error) {
      log.error('Cosmos:getAddressAndPublicKey', { error });

      return processErrorResponse(error);
    }
  }

  public async showAddressAndPubKey(
    path: number[] | string,
    hrp: string,
    p1: number = P1_VALUES.SHOW_ADDRESS_IN_DEVICE,
    p2: number = 0
  ): Promise<APK> {
    return this.getAddressAndPublicKey(path, hrp, p1, p2);
  }

  public async sign(
    path: number[] | string,
    data: any,
    size: number = SIZE
  ): Promise<Signature> {
    data = isArrayOrTypedArray(data)
      ? data
      : isObject(data)
      ? JSON.stringify(sortObject(data))
      : String(data);

    try {
      log.log('Cosmos:sign request', { path, data, size });

      const chunks: Buffer[] = await this.signGetChunks(path, data, size);

      log.log('Cosmos:sign response', { chunks });

      const signatures: Signature[] = [];

      for (let i: number = 0; i < chunks.length; i++) {
        signatures[i] = await this.signSendChunk(
          1 + i,
          chunks.length,
          chunks[i]
        );

        if (signatures[i].return_code !== ERROR_CODE.NoError)
          return signatures[i];
      }

      const signature: Signature = signatures[signatures.length - 1];

      log.log('Cosmos:sign', { signatures, signature });

      return signature;
    } catch (error) {
      log.error('Cosmos:sign', { error });

      return processErrorResponse(error);
    }
  }

  // --- PRIVATE ---------------------------------------------------------------

  private async send(
    cla: number,
    ins: number,
    p1: number,
    p2: number,
    data?: Buffer,
    statusList?: number[],
    ms?: number
  ): Promise<Uint8Array> {
    if (!ms || ms === Infinity) {
      return (await this.transport.send(
        cla,
        ins,
        p1,
        p2,
        data,
        statusList
      )) as Uint8Array;
    }

    return await Promise.race<Promise<Uint8Array>>([
      this.transport.send(cla, ins, p1, p2, data, statusList),
      wait(ms).then((): any => {
        throw {
          error_message: errorCodeToString(14),
          return_code: 14,
        };
      }),
    ]);
  }

  private async version(force: boolean = true): Promise<Version> {
    if (this?._version?.return_code === ERROR_CODE.NoError && !force) {
      return this._version;
    }

    try {
      log.log('getVersion request');

      const response: Uint8Array = await this.send(
        CLA,
        INS.GET_VERSION,
        0,
        0,
        undefined,
        undefined,
        10 * 1000
      );

      log.log('getVersion response', { response });

      const codes: Uint8Array = response.slice(-2);
      const code: number = codes[0] * 256 + codes[1];

      this._version = {
        device_locked: response[4] === 1,
        error_message: errorCodeToString(code),
        major: response[1],
        minor: response[2],
        patch: response[3],
        return_code: code,
        target_id: (response.length >= 9
          ? (response[5] << 24) +
            (response[6] << 16) +
            (response[7] << 8) +
            (response[8] << 0)
          : 0
        ).toString(16),
        test_mode: response[0] !== 0,
      };

      log.log('getVersion version', { version: this._version });

      return this._version;
    } catch (error) {
      log.error('getVersion', { error });

      return processErrorResponse(error);
    }
  }

  private async signGetChunks(
    path: number[] | string,
    data: string | Uint8Array | ReadonlyArray<number>,
    size: number
  ): Promise<Buffer[]> {
    const serializedPath: Buffer = await this.serializePath(path);

    const chunks: Buffer[] = [serializedPath];

    const buffer: Buffer = Buffer.from(data as any);

    // const buffer: Buffer = crypto.createHash('sha256').update(data as Uint8Array).digest()

    for (let i: number = 0; i < buffer.length; i += size) {
      let end: number = i + size;
      if (i > buffer.length) {
        end = buffer.length;
      }
      chunks.push(buffer.slice(i, end));
    }

    return chunks;
  }

  private async signSendChunk(
    chunkIdx: number,
    chunkNum: number,
    chunk: Buffer
  ): Promise<Signature> {
    const version: Version = await this.version();

    switch (version.major) {
      case 1:
        return signSendChunkv1(this.transport, chunkIdx, chunkNum, chunk);
      case 2:
        return signSendChunkv2(this.transport, chunkIdx, chunkNum, chunk);
      default:
        throw version;
    }
  }
}
